src/qutip_qip/vqa.py

```diff
@@ -0,0 +1,841 @@
+        blocks = [*self.blocks]
+        for _ in range(1, self.n_layers):
+            for block in list(filter(lambda b: not b.initial, self.blocks)):
+                blocks.append(block)
+        return blocks
+
+    def add_block(self, block):
+        """
+        Append a VQABlock instance to the circuit, and update the
+        user_gates dictionary if necessary.
+
+        Parameters
+        ----------
+        block: VQABlock
+        """
+        if not block.name:
+            block.name = "U" + str(len(self.blocks))
+        if block.name in list(map(lambda b: b.name, self.blocks)):
+            raise ValueError("Duplicate Block name in self.blocks")
+        self.blocks.append(block)
+        self.user_gates[block.name] = lambda angles=None: block.get_unitary(
+            angles
+        )
+
+    def get_free_parameters(self):
+        """
+        Compute the number of free parameters required
+        to evaluate the circuit.
+
+        Returns
+        -------
+        num_params: int
+            Number of free circuit parameters
+        """
+        initial_blocks = list(filter(lambda b: b.initial, self.blocks))
+        layer_blocks = list(filter(lambda b: not b.initial, self.blocks))
+
+        n_initial_params = sum(
+            list(map(lambda b: b.get_free_parameters(), initial_blocks))
+        )
+
+        n_layer_params = (
+            sum(list(map(lambda b: b.get_free_parameters(), layer_blocks)))
+            * self.n_layers
+        )
+
+        return n_initial_params + n_layer_params
+
+    def construct_circuit(self, angles):
+        """
+        Construct a circuit by specifying values for each
+        free parameter.
+
+        Parameters
+        ----------
+        angles: list of float
+            A list of dimension (n,) for n free parameters in the circuit
+
+        Returns
+        -------
+        circ: QuantumCircuit
+        """
+        circ = QubitCircuit(self.n_qubits)
+        circ.user_gates = self.user_gates
+        i = 0
+        for layer_num in range(self.n_layers):
+            for block in self.blocks:
+                if block.initial and layer_num > 0:
+                    continue
+                if block.is_native_gate:
+                    circ.add_gate(block.operator, targets=block.targets)
+                else:
+                    n = block.get_free_parameters()
+                    circ.add_gate(
+                        block.name,
+                        targets=list(range(self.n_qubits)),
+                        arg_value=angles[i : i + n] if n > 0 else None,
+                    )
+                    i += n
+        return circ
+
+    def get_initial_state(self):
+        """
+        Returns
+        -------
+        initial_state: Qobj
+            Initial circuit state
+        """
+        initial_state = basis(2, 0)
+        for _ in range(self.n_qubits - 1):
+            initial_state = tensor(initial_state, basis(2, 0))
+        return initial_state
```

Comment:
You're right - I think it's sensible to assume the initial state is 0, and the user can prepare a different initial state by adding `VQABlock`s with the "initial" flag set to true. Good call on the `basis([2] * n, [0] * n)`, that's much more convenient! 

I think it might be a bit more friendly to the user if they can call `VQA.get_initial_state()` rather than having to use `basis([2] * VQA.n_qubits, [0] * VQA.n_qubits)`, even if the function is that simple? Happy to hear your thoughts, though.

Reply:
Sure yeah it is a bit easier.